### DISCUSSION TITLE
invalid_request (Parameter redirect_uri does not match code)

### DISCUSSION THREAD

**Discussion by AdnanAliArshadtx** [panva/openid-client#531] [Q&A] [answered]
I am trying to get the authentication token  while this library but I am getting this error:
`invalid_request (Parameter redirect_uri does not match code)`

My code is below:
 ```
const issuer = await Issuer.Issuer.discover('https://login.issuer.de/openidsandbox');

  code_verifier = Issuer.generators.codeVerifier();
  code_challenge = Issuer.generators.codeChallenge(code_verifier);
  client = new issuer.Client({
    client_id: 'someclientid',
    client_secret: 'someclientsecret,
    redirect_uri: `http://localhost:3000/authorize/`,
    response_types: ['code id_token'],
    id_token_signing_alg_values_supported: "RS256",
  });
  
  url = client.authorizationUrl({
    // scope: 'openid functiontest',
    scope: 'openid',
    response_mode: 'form_post',
    nonce,
    code_challenge,
    code_challenge_method: 'S256',
    state,
  });
```
  
  It opens the authention page and after login  on server we are succesfully redirected back to our end point. which is handled by end-point  **authorize**
  
  ```
('/authorize/', async (request, reply) => {
  const params = client.callbackParams(request);
  const tokenSet = await client.callback('https://login.issuer.de/openidsandbox/authorize', params, { nonce, state });
  // this line throws the error.
  });
```
  
  Is there something I am doing wrong. Please suggest. 🙏

**Comment by panva** [comment, marked as answer]
The url passed to client.callback is supposed to be your redirect url. Not the issuer's authorization_endpoint.

**Reply by AdnanAliArshadtx** [reply]
I tried to pass my own server end-point but still same error.

**Reply by panva** [reply]
This value exactly. http://localhost:3000/authorize/

**Reply by AdnanAliArshadtx** [reply]
Thanks, you're right. It started throwing a different error `invalid_grant (Parameter code_verifier missing in token request)`
which I resolved by using line:
`const tokenSet = await client.callback('http://localhost:3000/authorize/', params, { nonce, state, code_verifier });`
and now I am at this error: **"Converting circular structure to JSON\n    --> starting at object with constructor 'Socket'\n    |     property 'parser' -> object with constructor 'HTTPParser'\n    --- property 'socket' closes the circle"**

**Reply by panva** [reply]
> Thanks, you're right. It started throwing a different error `invalid_grant (Parameter code_verifier missing in token request)` which I resolved by using line: `const tokenSet = await client.callback('http://localhost:3000/authorize/', params, { nonce, state, code_verifier });` and now I am at this error: **"Converting circular structure to JSON\n --> starting at object with constructor 'Socket'\n | property 'parser' -> object with constructor 'HTTPParser'\n --- property 'socket' closes the circle"**

That's no longer an error from openid-client.

**Reply by AdnanAliArshadtx** [reply]
Yes you're right and this was json error which has been fixed but overall fix was below line purposed by you.

`const tokenSet = await client.callback('http://localhost:3000/authorize/', params, { nonce, state, code_verifier });`
